nengo/utils/testing.py

```diff
@@ -184,14 +185,40 @@ def __exit__(self, type, value, traceback):
             del self.handler
 
 
+RecordedWarning = collections.namedtuple(
+    'RecordedWarning', ['message', 'category', 'filename', 'lineno', 'module'])
+
+
 class WarningCatcher(object):
+    def warn_explicit(self, message, category, filename, lineno,
+                      module=None, *args, **kwargs):
+        self.recorded.append(RecordedWarning(
+            message, category, filename, lineno, module))
+        self.old_warn_explicit(
+            message, category, filename, lineno, module, *args, **kwargs)
+
+    def warn(self, message, category=None, *args, **kwargs):
+        if isinstance(message, Warning):
+            category = type(message)
+        category = UserWarning if category is None else category
+        # warnings.warn uses sys._getframe and other mechanisms to get
+        # the filename, lineno, and module. We won't bother for now.
+        self.recorded.append(RecordedWarning(
+            message, category, None, None, None))
+        self.old_warn(message, category, *args, **kwargs)
+
     def __enter__(self):
-        self.catcher = warnings.catch_warnings(record=True)
-        self.record = self.catcher.__enter__()
-        warnings.simplefilter('always')
+        self.recorded = []
```

Comment:
If the same WarningCatcher is used in two if blocks, it will only retain the warnings of the last block. I would probably expect that all warnings will be retained. So, maybe move this line to `__init__`?